src/cmap/background.rs

```diff
@@ -53,13 +55,31 @@ impl ConnectionPool {
     /// each iteration and acquire it again during the next one to ensure that the this method
     /// doesn't block other threads from acquiring connections.
     async fn ensure_min_connections(&self) {
-        if let Some(min_pool_size) = self.inner.min_pool_size {
+        if let Some(min_pool_size) = self.min_pool_size {
             loop {
-                let mut connection_manager = self.inner.connection_manager.lock().await;
-                if connection_manager.total_connection_count < min_pool_size {
-                    match connection_manager.create_connection().await {
-                        Ok(connection) => {
-                            connection_manager.checked_in_connections.push(connection)
+                if self.total_connection_count.load(Ordering::SeqCst) < min_pool_size {
+                    // Reserve a spot via the wait queue. This will prevent too many threads from
+                    // concurrently creating connections such that max_pool_size is exceeded.
+                    let _wait_queue_handle = match self.wait_queue.try_skip_queue() {
+                        None => {
+                            // This branch is rarely taken because we _just_ verified that
+                            // total_connection_count < min_pool_size, which implies
+                            // total_connection_count <= max_pool_size. It is possible however to
+                            // take this branch if a bunch of operation threads enter the queue
+                            // between the count check and our attempt to reserve a spot. In that
+                            // case we just return early because those threads that did enter the
+                            // queue will ensure min_pool_size is fulfilled.
+                            return;
+                        }
+                        Some(handle) => handle,
+                    };
+
+                    let connection = self.create_pending_connection();
+                    match self.establish_connection(connection).await {
```

Comment:
this could be improved to be done in the background as well (with respect to other establishments required to meet minPoolSize), but I figured it probably didn't buy us much.